File: apps/api/src/auth/auth.service.ts
import {
  ConflictException,
  Injectable,
  UnauthorizedException,
} from "@nestjs/common";
import { createHash } from "node:crypto";
import { PrismaService } from "../common/prisma/prisma.service";
import { PasswordService } from "./password.service";
import { TokenService } from "./token.service";
import { OtpService } from "./otp.service";

function hashToken(token: string): string {
  return createHash("sha256").update(token).digest("hex");
}

export interface AuthResult {
  accessToken: string;
  refreshToken: string;
  user: {
    id: string;
    email: string | null;
    phone: string | null;
    roles: string[];
  };
}

@Injectable()
export class AuthService {
  constructor(
    private readonly prisma: PrismaService,
    private readonly passwordService: PasswordService,
    private readonly tokenService: TokenService,
    private readonly otpService: OtpService
  ) {}

  private async loadRoleNames(userId: string): Promise<string[]> {
    const userRoles = await this.prisma.userRole.findMany({
      where: { userId },
      include: { role: true },
    });
    return userRoles.map((ur: { role: { name: string } }) => ur.role.name);
  }

  private async issueSessionTokens(
    userId: string,
    email: string | null,
    roles: string[],
    ipAddress?: string,
    userAgent?: string
  ): Promise<{ accessToken: string; refreshToken: string }> {
    const accessToken = this.tokenService.signAccessToken({ sub: userId, email, roles });
    const { token: refreshToken, sessionId } = this.tokenService.signRefreshToken(userId);

    // A session row backs Part D's "device/session management" requirement
    // and lets an admin or the user revoke a single session without
    // invalidating every device (Part K tenant dashboard).
    //
    // Only a SHA-256 hash of the refresh token is stored, the same principle
    // as password hashing: a stolen database dump must not itself be usable
    // to impersonate a session, the way a stored plaintext token would be.
    await this.prisma.userSession.create({
      data: {
        id: sessionId,
        userId,
        refreshToken: hashToken(refreshToken),
        ipAddress,
        userAgent,
        expiresAt: new Date(Date.now() + 30 * 24 * 60 * 60 * 1000),
      },
    });

    return { accessToken, refreshToken };
  }

  private async logFailedLogin(email: string, ipAddress?: string): Promise<void> {
    // Never let the security-audit write itself fail the login flow, and
    // never store the attempted password — only the email and IP.
    await this.prisma.securityEvent
      .create({
        data: { eventType: "failed_login", ipAddress, metadataJson: { email } },
      })
      .catch(() => undefined);
  }

  async register(input: {
    email: string;
    phone: string;
    password: string;
    firstName?: string;
    lastName?: string;
  }): Promise<AuthResult> {
    const existing = await this.prisma.user.findFirst({
      where: { OR: [{ email: input.email }, { phone: input.phone }] },
    });
    if (existing) {
      throw new ConflictException("An account with this email or phone already exists.");
    }

    const passwordHash = await this.passwordService.hash(input.password);

    const tenantRole = await this.prisma.role.upsert({
      where: { name: "Tenant" },
      update: {},
      create: { name: "Tenant" },
    });

    const user = await this.prisma.user.create({
      data: {
        email: input.email,
        phone: input.phone,
        passwordHash,
        status: "PENDING_VERIFICATION",
        profile: {
          create: {
            firstName: input.firstName,
            lastName: input.lastName,
          },
        },
        roles: {
          create: { roleId: tenantRole.id },
        },
      },
    });

    const { accessToken, refreshToken } = await this.issueSessionTokens(user.id, user.email, [
      "Tenant",
    ]);

    return {
      accessToken,
      refreshToken,
      user: { id: user.id, email: user.email, phone: user.phone, roles: ["Tenant"] },
    };
  }

  async loginWithPassword(
    email: string,
    password: string,
    context?: { ipAddress?: string; userAgent?: string }
  ): Promise<AuthResult> {
    const user = await this.prisma.user.findUnique({ where: { email } });
    if (!user || !user.passwordHash) {
      await this.logFailedLogin(email, context?.ipAddress);
      throw new UnauthorizedException("Invalid email or password.");
    }

    const valid = await this.passwordService.verify(user.passwordHash, password);
    if (!valid) {
      await this.logFailedLogin(email, context?.ipAddress);
      throw new UnauthorizedException("Invalid email or password.");
    }

    if (user.status === "SUSPENDED") {
      throw new UnauthorizedException("This account has been suspended.");
    }

    const roles = await this.loadRoleNames(user.id);
    const { accessToken, refreshToken } = await this.issueSessionTokens(
      user.id,
      user.email,
      roles,
      context?.ipAddress,
      context?.userAgent
    );

    return {
      accessToken,
      refreshToken,
      user: { id: user.id, email: user.email, phone: user.phone, roles },
    };
  }

  async requestPhoneOtp(phone: string): Promise<{ expiresInSeconds: number }> {
    return this.otpService.generateAndSendOtp(phone);
  }

  async loginWithPhoneOtp(
    phone: string,
    code: string,
    context?: { ipAddress?: string; userAgent?: string }
  ): Promise<AuthResult> {
    const verified = await this.otpService.verifyOtp(phone, code);
    if (!verified) {
      throw new UnauthorizedException("Incorrect or expired code.");
    }

    let user = await this.prisma.user.findUnique({ where: { phone } });
    let roles: string[];

    if (!user) {
      const tenantRole = await this.prisma.role.upsert({
        where: { name: "Tenant" },
        update: {},
        create: { name: "Tenant" },
      });
      user = await this.prisma.user.create({
        data: {
          phone,
          phoneVerified: true,
          status: "ACTIVE",
          roles: { create: { roleId: tenantRole.id } },
        },
      });
      roles = ["Tenant"];
    } else {
      if (!user.phoneVerified) {
        await this.prisma.user.update({
          where: { id: user.id },
          data: { phoneVerified: true, status: "ACTIVE" },
        });
      }
      roles = await this.loadRoleNames(user.id);
    }

    const { accessToken, refreshToken } = await this.issueSessionTokens(
      user.id,
      user.email,
      roles,
      context?.ipAddress,
      context?.userAgent
    );

    return {
      accessToken,
      refreshToken,
      user: { id: user.id, email: user.email, phone: user.phone, roles },
    };
  }

  async refresh(refreshToken: string): Promise<AuthResult> {
    let payload;
    try {
      payload = this.tokenService.verifyRefreshToken(refreshToken);
    } catch {
      throw new UnauthorizedException("Invalid or expired refresh token.");
    }

    const session = await this.prisma.userSession.findUnique({
      where: { refreshToken: hashToken(refreshToken) },
    });
    if (!session || session.revokedAt || session.expiresAt < new Date()) {
      throw new UnauthorizedException("Session no longer valid. Please sign in again.");
    }

    const user = await this.prisma.user.findUnique({ where: { id: payload.sub } });
    if (!user) {
      throw new UnauthorizedException("User no longer exists.");
    }

    // Rotate the refresh token (revoke the old session, issue a new one) to
    // limit the blast radius of a leaked refresh token.
    await this.prisma.userSession.update({
      where: { id: session.id },
      data: { revokedAt: new Date() },
    });

    const roles = await this.loadRoleNames(user.id);
    const { accessToken, refreshToken: newRefreshToken } = await this.issueSessionTokens(
      user.id,
      user.email,
      roles
    );

    return {
      accessToken,
      refreshToken: newRefreshToken,
      user: { id: user.id, email: user.email, phone: user.phone, roles },
    };
  }

  async loginOrCreateFromGoogleProfile(profile: {
    email: string | null;
    firstName?: string;
    lastName?: string;
  }): Promise<AuthResult> {
    if (!profile.email) {
      throw new UnauthorizedException("Google account did not provide an email address.");
    }

    let user = await this.prisma.user.findUnique({ where: { email: profile.email } });
    let roles: string[];

    if (!user) {
      const tenantRole = await this.prisma.role.upsert({
        where: { name: "Tenant" },
        update: {},
        create: { name: "Tenant" },
      });
      user = await this.prisma.user.create({
        data: {
          email: profile.email,
          emailVerified: true,
          status: "ACTIVE",
          profile: {
            create: { firstName: profile.firstName, lastName: profile.lastName },
          },
          roles: { create: { roleId: tenantRole.id } },
        },
      });
      roles = ["Tenant"];
    } else {
      roles = await this.loadRoleNames(user.id);
    }

    const { accessToken, refreshToken } = await this.issueSessionTokens(
      user.id,
      user.email,
      roles
    );

    return {
      accessToken,
      refreshToken,
      user: { id: user.id, email: user.email, phone: user.phone, roles },
    };
  }

  async logout(refreshToken: string): Promise<void> {
    await this.prisma.userSession
      .updateMany({
        where: { refreshToken: hashToken(refreshToken), revokedAt: null },
        data: { revokedAt: new Date() },
      })
      .catch(() => undefined);
  }
}

File: apps/api-py/app/services/auth_service.py
"""
Direct port of apps/api/src/auth/auth.service.ts's register/login/refresh
flow. Kept as plain functions taking an AsyncSession, mirroring the
original's constructor-injected PrismaService — this makes it directly
unit-testable with a mocked session, the same way the original tests mock
PrismaService rather than hitting a real database.
"""

from datetime import datetime, timedelta, timezone

from fastapi import HTTPException, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.security import (
    hash_password,
    hash_token,
    sign_access_token,
    sign_refresh_token,
    verify_password,
)
from app.models import User, UserProfile, UserRole, Role, UserSession


async def load_role_names(db: AsyncSession, user_id: str) -> list[str]:
    result = await db.execute(
        select(Role.name).join(UserRole, UserRole.role_id == Role.id).where(UserRole.user_id == user_id)
    )
    return [row[0] for row in result.all()]


async def _issue_session_tokens(
    db: AsyncSession,
    user_id: str,
    email: str | None,
    roles: list[str],
    ip_address: str | None = None,
    user_agent: str | None = None,
) -> tuple[str, str]:
    access_token = sign_access_token(user_id, email, roles)
    refresh_token, session_id = sign_refresh_token(user_id)

    db.add(
        UserSession(
            id=session_id,
            user_id=user_id,
            refresh_token=hash_token(refresh_token),
            ip_address=ip_address,
            user_agent=user_agent,
            expires_at=datetime.now(timezone.utc) + timedelta(days=30),
        )
    )
    await db.commit()

    return access_token, refresh_token


async def register(
    db: AsyncSession,
    first_name: str,
    last_name: str,
    email: str,
    password: str,
    phone: str | None,
) -> tuple[str, str, User, list[str]]:
    existing = await db.execute(select(User).where(User.email == email))
    if existing.scalar_one_or_none() is not None:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="An account with this email already exists.")

    user = User(email=email, phone=phone, password_hash=hash_password(password))
    db.add(user)
    await db.flush()  # populate user.id before creating dependent rows

    db.add(UserProfile(user_id=user.id, first_name=first_name, last_name=last_name))

    # Every new registration is a Tenant by default — matches the original
    # ("Every registration always assigns 'Tenant' role").
    tenant_role = (await db.execute(select(Role).where(Role.name == "Tenant"))).scalar_one_or_none()
    if tenant_role is not None:
        db.add(UserRole(user_id=user.id, role_id=tenant_role.id))

    await db.commit()

    roles = await load_role_names(db, user.id)
    access_token, refresh_token = await _issue_session_tokens(db, user.id, user.email, roles)
    return access_token, refresh_token, user, roles


async def login(db: AsyncSession, email: str, password: str) -> tuple[str, str, User, list[str]]:
    result = await db.execute(select(User).where(User.email == email))
    user = result.scalar_one_or_none()

    # Same message whether the email doesn't exist or the password is
    # wrong — never reveal which one it was.
    invalid = HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Invalid email or password.")

    if user is None or user.password_hash is None:
        raise invalid
    if not verify_password(user.password_hash, password):
        raise invalid

    roles = await load_role_names(db, user.id)
    access_token, refresh_token = await _issue_session_tokens(db, user.id, user.email, roles)
    return access_token, refresh_token, user, roles
